src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py

```diff
@@ -88,52 +229,42 @@ def sync_engagement_db_to_rapid_pro(engagement_db, rapid_pro, uuid_table, sync_c
     _ensure_rapid_pro_has_contact_fields(rapid_pro, contact_fields_to_sync)
 
     # Load all the project code schemes so we can easily scan for STOP messages later.
-    code_schemes = {}
+    code_schemes = []
     for path in glob.glob("code_schemes/*.json"):
         with open(path) as f:
-            code_scheme = CodeScheme.from_firebase_map(json.load(f))
-            code_schemes[code_scheme.scheme_id] = code_scheme
+            code_schemes.append(CodeScheme.from_firebase_map(json.load(f)))
 
-    # Sync each participant to Rapid Pro.
-    for i, (participant_uuid, datasets) in enumerate(participants.items()):
-        log.info(f"Syncing participant {i + 1}/{len(participants)}: {participant_uuid}...")
-        # Re-identify the participant
-        urn = uuid_table.uuid_to_data(participant_uuid)
+    # Sync each message to Rapid Pro, by recomputing the state of every participant.
+    participants_synced_this_cycle = set()
+    for i, message in enumerate(messages_triggering_sync):
+        log.info(f"Syncing message {i + 1}/{len(messages_triggering_sync)}: {message.message_id}...")
+        participant_uuid = message.participant_uuid
+        if participant_uuid in participants_synced_this_cycle:
+            log.info(f"Skipping this message because we've already synced participant_uuid {participant_uuid} this run")
```

Comment:
```suggestion
            log.info(f"Skipping this message because we've already synced participant_uuid {participant_uuid} before this run")
```

Reply:
I think this is correct as-is because I'm using "run" to mean a complete run of this pipeline stage, and because the optimisation being made here is to fully sync each participant only once, rather than once per changed message.

Would it help to rephrase this log to we've already synced "in this pipeline run"?

Reply:
yes